client-src/elements/chromedash-app.js

```diff
@@ -138,99 +138,177 @@ class ChromedashApp extends LitElement {
   setUpRoutes() {
     page.strict(true); // Be precise about trailing slashes in routes.
 
-    const scrollIfHash = (ctx) => {
-      // If current page is ctx.path and only the hash changes, don't create a new element
-      // but instead just scroll to the element identified by the hash.
+    // Whether any changes have been made to form fields.
+    // Undo of changes does not undo this setting.
+    // This is null only when "loading" a new page.
+    let changesMade = null;
+
+    let beforeUnloadHandler = null;
+
+    const resetBeforeUnloadHandler = () => {
+      if (!beforeUnloadHandler) return;
+      // Remove previous beforeunload handler, to forget changes.
+      window.removeEventListener('beforeunload', beforeUnloadHandler);
+      beforeUnloadHandler = null;
+    };
+
+    // Maybe set up new page, or if the URL is the same, we stay.
+    // Returns true if we are proceeding to the new page, false otherwise.
+    const setupNewPage = (ctx, componentName) => {
+      // If current page is ctx.path and a ctx.hash exists,
+      // don't create a new element but instead
+      // just scroll to the element identified by the hash.
+      // Note, this ignores any query string.
       if (this.currentPage == ctx.path && ctx.hash) {
         if (window.scrollToElement) {
           window.scrollToElement(`#${ctx.hash}`);
         }
-        return true;
+        return false;
+      }
+
+      // If there was a previous page, check if we would lose unsaved changes.
+      if (this.pageComponent) {
+        // Act like we are unloading previous page and loading a new page.
+        if (changesMade) {
+          // Should we use shoelace dialog instead?
+          if (!confirm('You will lose unsaved changes.  Proceed anyway?')) {
+            // Set ctx.handled to false, so we don't change browser's history.
+            ctx.handled = false;
+            return false;
+          }
+        }
       }
+
+      // Loading new page.
+      this.pageComponent = document.createElement(componentName);
+      changesMade = null;
+      resetBeforeUnloadHandler();
```

Comment:
It seems overly complicated to register and unregister the same event handler on every page navigation.  Could we just register it once for the whole app and then store a boolean named `dirty` that is checked inside that event handler?

Reply:
Indeed it does seem overly complicated, but I'm not sure whether it is possible to register it once for the whole app.  There are a few complications: 

1. We only want the beforeunload event to show the dialog if there is an unsaved change to a form field.  We need to handle changes anyway, so setting changesMade (aka dirty) is consistent.  If we simply return from the beforeunload handler when there are not changesMade, that could work.
2. The Submit button (or 'submit' event?) needs to ignore the changes while the submit is happening.  I'm not sure I need to remove the beforeunload event handler to  make that happen; that's just what I saw others doing.  
3. Restoring the beforeunload event handler after a failure to successfully submit (for whatever reason) is the last complication.  There is no way, in general, to detect whether a failure might occur.  Invalid form values is one thing, but other failures are possible.
4. Once the beforeunload event handler is set, it seems form field changes are captured automatically, even those inside shadow DOM.  It's a bit mysterious exactly how that might be happening.  So when a "page" changes, the previous changes need to be forgotten, which also seems to require removing and restoring the beforeunload event handler. 

I'll do some more experimenting to see if I can simplify things a bit.  First priority was just getting it working.  After cleaning up some more, I'll look into whether we can test page changes including simulating the user responding to the beforeunload dialog.

Reply:
Seems to work fine to just return from the beforeunload handler if there are no changes (or during submit, if we are pretending there are no changes).  The overall logic is about the same, but we only have to set up the beforeunload handler one time.

Reply:
Perhaps I over-simplified the beforeunload handling.  

Regarding [Page Lifecycle API - Chrome Developers](https://developer.chrome.com/blog/page-lifecycle-api/#the-beforeunload-event)  Taymon Beal says:

> Per the above page, though, [permanent use of beforeunload] still breaks the bfcache, so don't install a permanent beforeunload handler. Instead, register a handler whenever the user makes a change, and then unregister it whenever you persist the current state, so that the handler only exists during those moments when there is unsaved work to warn the user about.

It's easy enough for me to switch back to doing that, but it can wait until I figure out how to test the inter-page event  transitions.